Updated Miniscript with timelock and heightlock information

sanket1729 opened the pull request:
Added functionality for checking combination of timelocks and height locks. 
Currently, I am focusing on trying to work on Miniscript without the compiler, but if required I can update the compiler to fail if given a policy that could contain such a combination. 

After going through all the remaining issues in the repo, I will adapt the PR to bitcoin core. 
Leaving some review comments for places I was not sure about the C++ idiomatic way to do things.

sanket1729 commented on bitcoin/script/miniscript.h:
```diff
@@ -293,6 +294,41 @@ struct StackSize {
+    //! csv with heights
+    bool csv_h;
+    //! csv with times
+    bool csv_t;
```
Oops, I will change this later for sure. Looks like github UI does `_t` in variable name for a good reason.

sipa replied:
Ending a name with `_t` is actually UB, I believe.

sanket1729 commented on bitcoin/script/miniscript.h:
```diff
@@ -568,6 +606,45 @@ struct Node {
+                for (const auto& sub : subs) {
+                    tl_info_vec.push_back(sub->tl_info);
+                }
+                return internal::TimeLockInfo(k, tl_info_vec);
```
This not seem ergonomic. The way I would do this in rust is to use a function that takes in an iterator as an argument. Is this the idiomatic way to consume a vector? 
In the current implementation, we iterate over the loop twice which is not ideal, but on the other hand, I also like that all the timelock boolean logic is inside one function for all fragments.

sipa replied:
```c++
std::vector<internal::TimeLockInfo> tl_info_vec;
std::transform(subs.begin(), subs.end(), std::back_inserter(tl_info_vec), [](const NodeRef<Key>& arg) { return arg->tl_info; });
return tl_info_vec;
```

perhaps?

sanket1729 commented on bitcoin/script/miniscript.h:
```diff
@@ -841,7 +844,7 @@ struct Node {
 
     //! Do all sanity checks.
-    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckOpsLimit() && CheckStackSize() && CheckTimeLocks(); }
+    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckScriptSize() && CheckOpsLimit() && CheckStackSize() && CheckTimeLocks(); }
```
I think we can change this as a validity check since this does not depend on the satisfactions and is known at parse time.

sipa commented on bitcoin/script/miniscript.h:
```diff
@@ -293,6 +294,41 @@ struct StackSize {
     StackSize(MaxInt<uint32_t> in_sat, MaxInt<uint32_t> in_dsat) : sat(in_sat), dsat(in_dsat) {};
 };
 
+struct TimeLockInfo {
```
Any reason why this is a separate structure, and not integrated with the other type properties?

sanket1729 replied:
When we calculate something as a type property, we get the benefit of caching intermediate nodes in the Miniscript struct. I didn't see any use in saving this information for intermediate miniscripts nodes, I did not make them type properties. 

We decided
1) The Miniscripts that contain a possible spend of timelock and heightlock are still well-typed valid miniscripts. 
2) For the policy compiler, we refuse to accept policies that can contain such a spend path
3) We don't do any semantic analysis on such Miniscripts because that can result in incorrect results. 

If you think I should create five more types in type-properties, I can change that too. I don't have any strong preference.

dgpv commented:
I've generated 1772 miniscripts with timelock conflicts, hopefully will be useful to test this code. The archive with `.dot` files is at https://github.com/dgpv/miniscript-alloy-spec-generated-samples/blob/master/8f1e80d95c49d49563b97fb97fd3ed38b55b4672_timelock_conflicts.zip, the code to convert from `.dot` file to the miniscript string representation is at https://github.com/dgpv/miniscript-alloy-spec/blob/master/tools/parse_miniscript_dot.py

darosior commented on bitcoin/script/miniscript.cpp:
```diff
@@ -28,7 +28,7 @@ Type SanitizeType(Type e) {
         (!(e << "V"_mst) ||  (e << "f"_mst)) && // V implies f
         (!(e << "K"_mst) ||  (e << "s"_mst)) && // K implies s
-        (!(e << "z"_mst) ||  (e << "m"_mst)); // z implies m
+        (!(e << "z"_mst) ||  (e << "m"_mst));// z implies m
```
Spurious change?

darosior commented on bitcoin/script/miniscript.h:
```diff
@@ -750,7 +750,10 @@ struct Node {
+    size_t GetScriptSize() const { return scriptlen; }
+
+    //! Check the size of the script against policy limits
+    bool CheckScriptSize() const { return GetScriptSize() <= MAX_STANDARD_P2WSH_SCRIPT_SIZE; }
```
You never call it:
```diff
diff --git a/bitcoin/script/miniscript.h b/bitcoin/script/miniscript.h
index 04855fb..b03dc39 100644
--- a/bitcoin/script/miniscript.h
+++ b/bitcoin/script/miniscript.h
@@ -786,7 +786,7 @@ public:
     bool NeedsSignature() const { return GetType() << "s"_mst; }
 
     //! Do all sanity checks.
-    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckOpsLimit() && CheckStackSize() && CheckTimeLocksMix(); }
+    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckOpsLimit() && CheckStackSize() && CheckTimeLocksMix() && CheckScriptSize(); }
 
     //! Construct the script for this miniscript (including subexpressions).
     template<typename Ctx>
```

sanket1729 replied:
Used the function in compiler checks. These functions would be used again when integrating with bitcoin core descriptors.

darosior replied:
Cool. I still think that it should be included to `IsSafeTopLevel` too though?

sanket1729 replied:
It is right? In the same diff that you included?

darosior replied:
`git fetch sanket && git checkout sanket/timelocks && git grep CheckScriptSize` says otherwise :) My diff was to ask for you to include it

sanket1729 replied:
Thanks, good catch :). I had many miniscript trees locally and I messed this up.

darosior commented on main.cpp:
```diff
@@ -19,9 +19,9 @@ static bool run(std::string&& line, int64_t count) {
+        printf("X %17.10f %5i %s %s\n", ret->GetScriptSize() + avgcost, (int)ret->GetScriptSize(), Abbreviate(std::move(str)).c_str(), line.c_str());
     } else if ((ret = miniscript::FromString(Expand(line), COMPILER_CTX))) {
-        printf("%7li scriptlen=%i maxops=%i type=%s safe=%s nonmal=%s dissat=%s input=%s output=%s miniscript=%s\n", (long)count, (int)ret->ScriptSize(), (int)ret->GetOps(), ret->GetType() << "B"_mst ? "B" : ret->GetType() << "V"_mst ? "V" : ret->GetType() << "W"_mst ? "W" : ret->GetType() << "K"_mst ? "K" : "(invalid)", ret->GetType() << "s"_mst ? "yes" : "no", ret->GetType() << "m"_mst ? "yes"  ...
+        printf("%7li scriptlen=%i maxops=%i type=%s safe=%s nonmal=%s dissat=%s input=%s output=%s timelock_mix=%s miniscript=%s\n", (long)count, (int)ret->GetScriptSize(), (int)ret->GetOps(), ret->GetType() << "B"_mst ? "B" : ret->GetType() << "V"_mst ? "V" : ret->GetType() << "W"_mst ? "W" : ret->GetType() << "K"_mst ? "K" : "(invalid)", ret->GetType() << "s"_mst ? "yes" : "no", ret->GetType()  ...
```
nit: the `timelock_mix` belongs to the previous commit

sipa commented on bitcoin/script/miniscript.h:
```diff
@@ -148,6 +162,11 @@ inline constexpr Type operator"" _mst(const char* c, size_t l) {
+        *c == 'h' ? 1 << 15 : // older: contains relative height timelock (csv_height)
+        *c == 'i' ? 1 << 16 : // after: contains time timelock   (cltv_time)
+        *c == 'j' ? 1 << 17 : // after: contains height timelock   (cltv_height)
+        *c == 'k' ? 1 << 18 : // contains a combination of height and time
```
I feel that using the negation of this (does not contain a combination of height and time) would be more in line with the existing properties (which are also "positive" ones, like safeness/nonmalleability). Then the compiler can be requested to select for that property, rather than needing to exclude it separately.

sanket1729 replied:
Agreed

sanket1729 replied:
Done

darosior commented on compiler.cpp:
```diff
@@ -402,14 +402,13 @@ struct Compilation {
         double cost = Cost(pair, node);
-        if (!node->CheckOpsLimit()) return;
-        if (node->GetStackSize() > MAX_STANDARD_P2WSH_STACK_ITEMS) return;
+        if (!node->CheckOpsLimit() || !node->CheckScriptSize() || !node->CheckStackSize()) return;
```
I was going to comment that you should add `CheckTimeLocksMix()` here too, but actually let's just replace everything by `IsSafeTopLevel()` to avoid the redundancy?

sanket1729 replied:
Done

sanket1729 commented on bitcoin/script/miniscript.cpp:
```diff
@@ -170,27 +202,44 @@ Type ComputeType(NodeType nodetype, Type x, Type y, Type z, const std::vector<Ty
+                    // Thresh contains a combination of timelocks if it has threshold > 1 and
+                    // it contains two different children that have different types of timelocks
+                    // Note how if any of the children don't have "k", the parent also does not have "k"
+                    "k"_mst.If(((acc_tl & t) << "k"_mst) && ((k <= 1) ||
```
This might be a bit tricky to review. Let me know if there is some better expression for this.

sanket1729 commented:
@dgpv, I looked at the tests but it looks like there is something off about the timelock conflict detection algorithm. 

The current suggested algorithm only makes sure that the canonical satisfaction paths don't have any conflicts. So, in the miniscript `or_i(after(1), after(500000001))` would not be any conflict because the user should not take both paths together. 

On the other hand, `and_b(after(1),a:after(500000001))` would be a conflict since there is a satisfaction path that requires both types of timelocks.

sipa commented:
@sanket1729 Thinking a bit more about this height/timelock conflict... is it actually a property of a miniscript, or of the policy? In other words: are there policies that can be compiled to both non-conflicting and conflicting forms?

sanket1729 commented:
@sipa  This is a property of the policy. There are no policies that can be compiled to both non-conflicting and conflicting forms. 

Our transcript from last year.
```
To summarize, we decided the following w.r.t to timelock conflicts.

- The Miniscripts that contain a possible spend of timelock and height locks are still well-typed valid miniscripts.
- For the policy compiler, we refuse to accept policies that can contain such a spend path
- We don't do any semantic analysis on such Miniscripts because that can result in incorrect results.
```

There is also safety/hassig `s` property which only depends on the policy. Our rationale for making it a miniscript property is it should be easy for software to detect such things without relying on the policy part.

sipa commented:
@sanket1729 Ok, thanks, that makes sense.

sanket1729 commented:
> I've generated 1772 miniscripts with timelock conflicts, hopefully will be useful to test this code. The archive with .dot files is at https://github.com/dgpv/miniscript-alloy-spec-generated-samples/blob/master/8f1e80d95c49d49563b97fb97fd3ed38b55b4672_timelock_conflicts.zip, the code to convert from .dot file to the miniscript string representation is at https://github.com/dgpv/miniscript-alloy-spec/blob/master/tools/parse_miniscript_dot.py

Thanks a lot. I have updated the rust-miniscript and c++ codebase to use those vectors. As mentioned earlier (https://github.com/sipa/miniscript/pull/41#issuecomment-885823488), not all of them are conflicts. But the c++ and rust-miniscript versions agree on those vectors. 

Because tests in the current master are based on an old version of miniscript, I have added tests in https://github.com/sipa/miniscript/pull/58/ which first updates the tests to the latest miniscript spec, and then adds new tests.

dgpv commented:
@sanket1729 The conditions I've chosen to generate those cases were apparently too simplistic, and now I even forgot the conditions I used to generate these, so just filtering this data is the easier path :-). You already filtered out the 'not really conflicts' cases, so all is fine I suppose.

darosior reviewed (approved):
ACK 7bba0e9076592aa1b691b8fb6aabd0775ffc1410

sipa commented on bitcoin/script/miniscript.h:
```diff
@@ -750,7 +750,10 @@ struct Node {
 
 public:
     //! Return the size of the script for this expression (faster than ToString().size()).
```
Unrelated: this should be `ToScript().size()`.

sipa commented on bitcoin/script/miniscript.h:
```diff
@@ -780,7 +783,7 @@ struct Node {
 
     //! Do all sanity checks.
-    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckOpsLimit() && CheckStackSize(); }
+    bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && CheckOpsLimit() && CheckStackSize() && CheckTimeLocksMix(); }
```
Now that timelock-safeness is just the "k" property, this could be simplified to just changing the required type to be `"Bmsk"_mst`. If `CheckTimeLocksMix` is independently interesting, it doesn't matter much.

sipa commented on compiler.cpp:
```diff
@@ -402,14 +402,13 @@ struct Compilation {
         double cost = Cost(pair, node);
-        if (!node->CheckOpsLimit()) return;
-        if (node->GetStackSize() > MAX_STANDARD_P2WSH_STACK_ITEMS) return;
+        if (!node->IsSafeTopLevel()) return;
```
I don't think this is right; it will reject all non-B expressions.

darosior replied:
We could have both a `IsSafe` and `IsSafeTopLevel` and call the former here?
```cpp
bool IsSafe() const { return CheckOpsLimit() && CheckStackSize() && CheckTimeLocksMix() && CheckScriptSize(); }

bool IsSafeTopLevel() const { return GetType() << "Bms"_mst && IsSafe(); }
```

sipa replied:
I think "ms" should be required for `IsSafe`, not just `IsSafeTopLevel`.

sipa reviewed (commented):
Feel like adding this to the website interface (js wrapper) too?

sanket1729 commented:
> Feel like adding this to the website interface (js wrapper) too?
   
   Yes, I will add another commit to this PR

sanket1729 commented:
Rebased and removed all the resource limitations checks for `MAX_WITNESS_SCRIPT_SIZE` from this PR. Those are being dealt with separate PRs by Darosior.

sipa commented on index.html:
```diff
@@ -411,6 +411,14 @@ <h4>Correctness properties</h4>
+The <samp>nSequence</samp> field in a transaction input, or the <samp>nLockTime</samp> field in transaction can be specified either as a time or height but not both.
+Therefore it is not possible to spend scripts that require satisfaction of both, height based timelock and time based timelock of the same type.
+<ul>
+<li><b>"k"</b> No timelock mixing. This expression does not contain a mix of heightlock and timelock of the same type. If the miniscript does not have the "k" property, the miniscript template will not match the user expectation of the corresponding spending policy.</li>
```
Mention the other types here too, or alternatively, drop them from `Props()` below?

sanket1729 replied:
Dropped from `Props()` as other properties are just implementation detail to obtain `k`

darosior replied:
nit: shouldn't this be part of the above list with the other properties? (`"z"`, `"o"`, `"n"`, `"d"`, `"u"`)

sanket1729 commented:
I have a few tests that I can push, but currently those would conflict with #76. I can push those once we resolve my comment in #73

darosior reviewed (commented):
ACK 1838d3e8d04c48df8e4d33ff34af174545d9c1bc

Verified especially that the `ComputeType` diff didn't change since my last review (and therefore i didn't re-check it).

darosior commented:
FWIW i have a rebased #75 on top of this so sounds better to go with this PR first (as it's more complex to both review and rebase): https://github.com/darosior/miniscript/tree/rebased_is_sane

sipa commented:
ACK 1838d3e8d04c48df8e4d33ff34af174545d9c1bc
